File: NET/Libraries/Treehopper.Libraries/Displays/Tm1650.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Treehopper.Libraries.Displays
{
    /// <summary>
    /// Titan Micro TM1650 4x8 LED driver
    /// </summary>
    /// <remarks>
    /// <para>Instead of using standard SMBus-style register/value pairs, this IC's individual registers are all accessed directly from different I2C addresses. Consequently, be aware that this IC uses quite a few I2C addresses — 0x24, 0x25, 0x26, 0x27 (for control), and 0x34, 0x35, 0x36, and 0x37 (for display).</para>
    /// </remarks>
    public class Tm1650 : LedDriver
    {
        private readonly I2c i2c;

        private static readonly byte ControlBase = 0x24;
        private static readonly byte DisplayBase = 0x34;

        private readonly byte[] oldValues = new byte[4];
        private readonly byte[] newValues = new byte[4];
        private bool enable;

        /// <summary>
        /// Construct a new TM1650 with a given I2c interface
        /// </summary>
        /// <param name="i2c">The I2c interface to use</param>
        public Tm1650(I2c i2c) : base(32, true, false)
        {
            this.i2c = i2c;
            brightness = 1.0;
            i2c.Enabled = true;
            Enable = true;
        }

        /// <summary>
        /// Enable or disable the display
        /// </summary>
        public bool Enable
        {
            get { return enable; }
            set
            {
                if (enable == value) return;
                enable = value;
                sendControlUpdate().Wait();
            }
        }

        private async Task sendControlUpdate()
        {
            int bright = (int)(Brightness < 1.0 ? Math.Ceiling(Brightness * 7.0) : 0);
            byte controlByte = (byte)((bright << 4) | ((enable && Brightness > 0) ? 0x01 : 0x00));
            for (int i = 0; i < 4; i++)
                await sendControl(controlByte, i);
        }

        private Task sendControl(byte data, int digit)
        {
            return i2c.SendReceive((byte)(ControlBase + digit), new byte[] { data }, 0);
        }

        private Task sendDisplay(byte data, int digit)
        {
            return i2c.SendReceive((byte)(DisplayBase + digit), new byte[] { data }, 0);
        }

        /// <summary>
        /// Flush data to the driver
        /// </summary>
        /// <param name="force">Whether or not to force all data to be flushed, even if it doesn't appear to have changed</param>
        /// <returns>An awaitable task that completes when finished</returns>
        public override async Task Flush(bool force = false)
        {
            for(int i=0;i<4;i++)
            {
                if (oldValues[i] != newValues[i] || force)
                {
                    await sendDisplay(newValues[i], i);
                    oldValues[i] = newValues[i];
                }
            }
        }

        internal override void LedBrightnessChanged(Led led)
        {
            
        }

        internal override void LedStateChanged(Led led)
        {
            int digit = led.Channel / 8;
            int segment = led.Channel % 8;
            bool value = led.State;

            // set or clear the appropriate bit
            if(led.State)
                newValues[digit] |= (byte)(1 << segment);
            else
                newValues[digit] &= (byte)~(1 << segment);

            if (AutoFlush)
                Flush().Wait();
        }

        internal override void SetGlobalBrightness(double brightness)
        {
            sendControlUpdate().Wait();
        }
    }
}

File: Java/treehopper.libraries/src/main/java/io/treehopper/libraries/displays/Tm1650.java
package io.treehopper.libraries.displays;

import io.treehopper.*;

/**
 * Created by jay on 12/5/2016.
 */

public class Tm1650 extends LedDriver {
    private I2c i2c;
    private static final byte ControlBase = 0x24;
    private static final byte DisplayBase = 0x34;
    private byte[] oldValues = new byte[4];
    private byte[] newValues = new byte[4];

    public boolean isEnabled() {
        return enable;
    }

    public void setEnabled(boolean enable) {
        if(this.enable == enable) return;
        this.enable = enable;
        sendControlUpdate();
    }

    private boolean enable;

    public Tm1650(I2c i2c)
    {
        super(32, false, false);
        this.i2c = i2c;
        i2c.setEnabled(true);
        this.enable = true;
        this.clear();
    }

    private void sendControlUpdate()
    {
        byte controlByte = (byte)(enable ? 0x01 : 0x00);
        for (int i = 0; i < 4; i++)
            sendControl(controlByte, i);
    }

    private void sendControl(byte data, int digit)
    {
        i2c.sendReceive((byte)(ControlBase + digit), new byte[] { data }, 0);
    }

    private void sendDisplay(byte data, int digit)
    {
        i2c.sendReceive((byte)(DisplayBase + digit), new byte[] { data }, 0);
    }

    @Override
    public boolean hasGlobalBrightnessControl() {
        return true;
    }

    @Override
    public boolean hasIndividualBrightnessControl() {
        return false;
    }

    @Override
    public double getBrightness() {
        return 0;
    }

    @Override
    public void setBrightness(double brightness) {

    }

    @Override
    public void flush(boolean force)
    {
        for(int i=0;i<4;i++)
        {
            if (oldValues[i] != newValues[i] || force)
            {
                sendDisplay(newValues[i], i);
                oldValues[i] = newValues[i];
            }
        }
    }

    public void ledBrightnessChanged(Led led)
    {

    }

    public void ledStateChanged(Led led)
    {
        int digit = led.getChannel() / 8;
        int segment = led.getChannel() % 8;
        boolean value = led.getState();

        // set or clear the appropriate bit
        if(led.getState())
            newValues[digit] |= (byte)(1 << segment);
        else
            newValues[digit] &= (byte)~(1 << segment);

        if (isAutoFlushEnabled())
            flush(false);
    }

}
